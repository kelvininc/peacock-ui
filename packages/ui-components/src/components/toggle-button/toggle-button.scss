@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --button-width: toggle button component's width.
	 * @prop --button-height: toggle button component's height.
	 * @prop --button-icon-width: toggle button component's icon width.
	 * @prop --button-icon-height: toggle button component's icon height.
	 * @prop --button-padding: toggle button component's padding.
	 * @prop --text-color-default: toggle button component's text color default.
	 * @prop --text-color-active: toggle button component's text color active.
	 * @prop --text-color-disabled: toggle button component's text color disabled.
	 * @prop --background-color-default: toggle button component's background color default.
	 * @prop --background-color-active: toggle button component's background color active.
	 * @prop --background-color-disabled: toggle button component's background color disabled.
	 * @prop --border-color-default: toggle button component's border color default.
	 * @prop --border-color-active: toggle button component's border color active.
	 * @prop --border-color-disabled: toggle button component's border color disabled.
	 */

	--text-color-default: #{kv-color('neutral-2')};
	--text-color-active: #{kv-color('neutral-0')};
	--text-color-disabled: #{kv-color('neutral-5')};
	--background-color-default: #{kv-color('neutral-6')};
	--background-color-active: #{kv-color('neutral-6')};
	--background-color-disabled: #{kv-color('neutral-7')};
	--border-color-default: #{kv-color('neutral-6')};
	--border-color-active: #{kv-color('neutral-0')};
	--border-color-disabled: #{kv-color('neutral-6')};
}

.toggle-button {
	@include kv-font-label-small-regular;

	text-decoration: none;

	position: relative;
	user-select: none;
	cursor: pointer;
	box-sizing: border-box;

	display: inline-flex;
	align-items: center;
	justify-content: center;
	gap: $spacing-2x;

	border-radius: 2px;

	border: 1px solid var(--border-color-default);
	color: var(--text-color-default);
	background: var(--background-color-default);

	width: var(--button-width, initial);
	height: var(--button-height, 36px);
	padding: var(--button-padding, 0 $spacing-3x);

	.toggle-button-icon,
	.toggle-button-icon kv-icon,
	.toggle-button-label {
		display: flex;
		align-items: center;
	}

	&--checked {
		@include kv-font-label-small-semibold;

		border-color: var(--border-color-active);
		color: var(--text-color-active);
		background: var(--background-color-active);

		.toggle-button-icon kv-icon {
			--icon-color: var(--text-color-active);
		}
	}

	&--disabled {
		border-color: var(--border-color-disabled);
		color: var(--text-color-disabled);
		background: var(--background-color-disabled);

		cursor: not-allowed;
		pointer-events: none;

		.toggle-button-icon kv-icon {
			--icon-color: var(--text-color-disabled);
		}
	}

	&--only-icon {
		width: var(--button-width, 28px);
		height: var(--button-height, 28px);
		padding: var(--button-padding, $spacing);

		kv-icon {
			--icon-color: var(--text-color-default);
			--icon-width: var(--button-icon-width, 20px);
			--icon-height: var(--button-icon-height, 20px);
			width: var(--button-icon-width, 20px);
			height: var(--button-icon-height, 20px);
		}
	}
}
