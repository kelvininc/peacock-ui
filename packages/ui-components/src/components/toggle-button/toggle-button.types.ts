import { EventEmitter } from '@stencil/core';
import { IAnchor } from '../../types';
import { EIconName, EOtherIconName } from '../icon/icon.types';

export interface IToggleButton extends IAnchor {
	/** (required) The value to be emitted upon click events */
	value: string;
	/** (optional) The button's icon. Only valid for toggle button icon */
	icon?: EIconName | EOtherIconName;
	/** (optional) The button's label. Only valid for toggle button text */
	label?: string;
	/** (optional) Sets the button's styling to be disabled and disables click events */
	disabled?: boolean;
	/** (optional) Sets the button as checked */
	checked?: boolean;
	/** (optional) Defines if the item click event should prevent default behaviour. */
	preventDefault?: boolean;
	/** (optional) Sets if the button is a radio button */
	withRadio?: boolean;
}

export interface IToggleButtonEvents {
	/** Emits when a button is clicked */
	checkedChange: EventEmitter<string>;
}
