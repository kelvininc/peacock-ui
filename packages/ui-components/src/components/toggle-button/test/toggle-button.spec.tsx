import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvToggleButton } from '../toggle-button';

describe('Toggle Button (unit tests)', () => {
	let page: SpecPage;
	let component: KvToggleButton;

	describe('when the component loads with label', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleButton],
				html: '<kv-toggle-button label="Option 1" value="opt1"></kv-toggle-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with icon', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleButton],
				html: '<kv-toggle-button icon="kv-add" value="opt1"></kv-toggle-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with label and icon', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleButton],
				html: '<kv-toggle-button label="Add item" icon="kv-add" value="opt1"></kv-toggle-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with disabled prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleButton],
				html: '<kv-toggle-button label="Option 1" value="opt1" disabled></kv-toggle-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with checked prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleButton],
				html: '<kv-toggle-button label="Option 1" value="opt1" checked></kv-toggle-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when the component loads with preventDefault prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleButton],
				html: '<kv-toggle-button label="Option 1" value="opt1" prevent-default="true"></kv-toggle-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		describe('and onClick is called', () => {
			let onClickSpyEvent: MouseEvent;

			beforeEach(() => {
				onClickSpyEvent = new MouseEvent('click');
				component.onClick(onClickSpyEvent);
			});

			it('should prevent event', () => {
				expect(onClickSpyEvent.defaultPrevented).toBeTruthy();
			});
		});
	});
});
