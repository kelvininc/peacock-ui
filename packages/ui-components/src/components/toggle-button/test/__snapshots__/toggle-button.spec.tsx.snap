// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Button (unit tests) when the component loads with checked prop should match the snapshot 1`] = `
<kv-toggle-button checked="" label="Option 1" value="opt1">
  <mock:shadow-root>
    <a class="toggle-button toggle-button--checked" part="toggle-button">
      <div class="toggle-button-label" part="toggle-label">
        Option 1
      </div>
    </a>
  </mock:shadow-root>
</kv-toggle-button>
`;

exports[`Toggle Button (unit tests) when the component loads with disabled prop should match the snapshot 1`] = `
<kv-toggle-button disabled="" label="Option 1" value="opt1">
  <mock:shadow-root>
    <a class="toggle-button toggle-button--disabled" part="toggle-button">
      <div class="toggle-button-label" part="toggle-label">
        Option 1
      </div>
    </a>
  </mock:shadow-root>
</kv-toggle-button>
`;

exports[`Toggle Button (unit tests) when the component loads with icon should match the snapshot 1`] = `
<kv-toggle-button icon="kv-add" value="opt1">
  <mock:shadow-root>
    <a class="toggle-button toggle-button--only-icon" part="toggle-button">
      <div class="toggle-button-icon" part="toggle-icon">
        <kv-icon name="kv-add"></kv-icon>
      </div>
    </a>
  </mock:shadow-root>
</kv-toggle-button>
`;

exports[`Toggle Button (unit tests) when the component loads with label and icon should match the snapshot 1`] = `
<kv-toggle-button icon="kv-add" label="Add item" value="opt1">
  <mock:shadow-root>
    <a class="toggle-button" part="toggle-button">
      <div class="toggle-button-icon" part="toggle-icon">
        <kv-icon name="kv-add"></kv-icon>
      </div>
      <div class="toggle-button-label" part="toggle-label">
        Add item
      </div>
    </a>
  </mock:shadow-root>
</kv-toggle-button>
`;

exports[`Toggle Button (unit tests) when the component loads with label should match the snapshot 1`] = `
<kv-toggle-button label="Option 1" value="opt1">
  <mock:shadow-root>
    <a class="toggle-button" part="toggle-button">
      <div class="toggle-button-label" part="toggle-label">
        Option 1
      </div>
    </a>
  </mock:shadow-root>
</kv-toggle-button>
`;

exports[`Toggle Button (unit tests) when the component loads with preventDefault prop should match the snapshot 1`] = `
<kv-toggle-button label="Option 1" prevent-default="true" value="opt1">
  <mock:shadow-root>
    <a class="toggle-button" part="toggle-button">
      <div class="toggle-button-label" part="toggle-label">
        Option 1
      </div>
    </a>
  </mock:shadow-root>
</kv-toggle-button>
`;
